security/s2a/internal/crypter/counter_test.go

```diff
@@ -0,0 +1,105 @@
+package crypter
+
+import (
+	"math"
+	"testing"
+)
+
+// counterFromValue creates a new counter given an initial value.
+func counterFromValue(value uint64) counter {
+	c := newCounter()
+	c.value = value
+	return c
+}
+
+func TestCounterInc(t *testing.T) {
+	for _, tc := range []struct {
+		desc                     string
+		counter, expectedCounter uint64
+		overflow                 bool
+	}{
+		{
+			desc:            "basic 1",
+			counter:         0,
+			expectedCounter: 1,
+		},
+		{
+			desc:            "basic 2",
+			counter:         123,
+			expectedCounter: 124,
+		},
+		{
+			desc:            "almost overflow",
+			counter:         math.MaxUint64 - 1,
+			expectedCounter: math.MaxUint64,
+		},
+		{
+			desc:     "max overflow",
+			counter:  math.MaxUint64,
+			overflow: true,
+		},
+	} {
+		t.Run(tc.desc, func(t *testing.T) {
+			c := counterFromValue(tc.counter)
+			c.inc()
+			val, err := c.val()
+			if tc.overflow {
+				if err == nil {
+					t.Errorf("counter starting with %v, val() expected error, received none", tc.counter)
+				}
+			} else {
+				if err != nil {
+					t.Errorf("counter starting with %v, val() returned error: %v", tc.counter, err)
+				}
+				if val != tc.expectedCounter {
+					t.Errorf("counter starting with %v, val() = %v, want %v", tc.counter, val, tc.expectedCounter)
+				}
+			}
+
+			if tc.overflow != c.hasOverflowed {
+				t.Errorf("counter starting with %v, c.hasOverflowed = %v, want %v", tc.counter, c.hasOverflowed, tc.overflow)
+			}
+		})
+	}
+}
+
+func TestCounterReset(t *testing.T) {
+	for _, tc := range []struct {
+		desc          string
+		counter       uint64
+		hasOverflowed bool
+	}{
+		{
+			desc:          "non-zero no overflow",
+			counter:       1,
+			hasOverflowed: false,
+		},
+		{
+			desc:          "zero no overflow",
+			counter:       0,
+			hasOverflowed: false,
+		},
+		{
+			desc:          "non-zero has overflow",
+			counter:       1,
+			hasOverflowed: true,
+		},
+		{
+			desc:          "zero has overflow",
+			counter:       0,
+			hasOverflowed: true,
+		},
+	} {
+		t.Run(tc.desc, func(t *testing.T) {
+			c := counter{tc.counter, tc.hasOverflowed}
+			// Check that resetting the counter works as expected.
```

Comment:
No need for this comment since this is implied from the fact that this test exists :)

Reply:
Done